src/theory/uf/theory_uf.cpp

```diff
@@ -224,6 +226,18 @@ TrustNode TheoryUF::ppRewrite(TNode node, std::vector<SkolemLemma>& lems)
       return TrustNode::mkTrustRewrite(node, ret, nullptr);
     }
   }
+  else if (k == kind::APPLY_UF)
```

Comment:
When can it happen that we have both cases? And -- do we still need `HO_APPLY`?

Reply:
Both are needed.  APPLY_UF is for fully applied function applications, HO_APPLY is for partially applied function applications. Sometimes a fully applied function term needs to be equated to an HO_APPLY term as well.  This gets into the details of this paper: https://homepage.divms.uiowa.edu/~ajreynol/cade19b.pdf

Reply:
But if we lift the requirement of having APPLY_UF with higher-order stuff we could do away with HO_APPLY, no?

Reply:
(Not that it should be done in this PR...)

Reply:
These two issues are not related actually.

That has to do with partial vs total applications.  This PR has to do with functions as arguments.

Reply:
Can we have a term with APPLY_UF have a functional return type? If we can't then I see why it's still necessary to have HO_APPLY.

Reply:
It's true that this would eliminate the first barrier for not needing HO_APPLY. But regardless, the reason we need HO_APPLY is because of function equality.  We won't do congruence over equality between APPLY_UF *operators*. Unless there is yet more magic in the equality engine we don't realize. :)

Reply:
haha yes, I think the big motivation is practicality, but I wanted to be sure that we don't need them, we just want them. :)